src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+
+                // GroupClause is not supported by the conversion
+                if (queryBody.SelectOrGroup.Kind() != SyntaxKind.SelectClause)
+                {
+                    return false;
+                }
+
+                queryExpressionProcessingInfo.Add(queryBody.SelectOrGroup);
+                return queryBody.Continuation == null || TryProcessQueryBody(queryBody.Continuation.Body, queryExpressionProcessingInfo);
+            }
+
+            private static BlockSyntax AddToBlockTop(StatementSyntax newStatement, StatementSyntax statement)
+            {
+                if (statement is BlockSyntax block)
+                {
+                    return block.WithStatements(block.Statements.Insert(0, newStatement));
+                }
+                else
+                {
+                    return SyntaxFactory.Block(newStatement, statement);
+                }
+            }
+
+            private bool IsLocalOrParameterSymbol(IOperation operation)
+            {
+                if (operation is IConversionOperation conversion && conversion.IsImplicit)
+                {
+                    return IsLocalOrParameterSymbol(conversion.Operand);
+                }
+
+                return operation.Kind == OperationKind.LocalReference || operation.Kind == OperationKind.ParameterReference;
+            }
+
+            private static BlockSyntax WrapWithBlock(StatementSyntax statement)
+                => statement is BlockSyntax block ? block : SyntaxFactory.Block(statement);
+
+            // Checks if the node is within an immediate lambda or within an immediate anonymous method.
+            // 'lambda => node' returns true
+            // 'lambda => localfunction => node' returns false
+            // 'member => node' returns false
+            private static bool IsWithinImmediateLambdaOrAnonymousMethod(SyntaxNode node)
+            {
+                while (node != null)
+                {
+                    switch (node.Kind())
+                    {
+                        case SyntaxKind.AnonymousMethodExpression:
+                        case SyntaxKind.ParenthesizedLambdaExpression:
+                        case SyntaxKind.SimpleLambdaExpression:
+                            return true;
+                        case SyntaxKind.LocalFunctionStatement:
+                            return false;
+                        default:
+                            if (node is MemberDeclarationSyntax)
+                            {
+                                return false;
+                            }
+
+                            break;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return false;
+            }
+
+            private class QueryExpressionProcessingInfo
+            {
+                public Stack<CSharpSyntaxNode> Stack { get; private set; }
+
+                public List<SyntaxToken> Identifiers { get; private set; }
```

Comment:
>public List<SyntaxToken> Identifiers { get; private set; } [](start = 16, length = 58)

It feels like a HashSet is a better fit for this member.

Reply:
I mean a HashSet of strings

---
In reply to: [178336137](https://github.com/dotnet/roslyn/pull/25362#discussion_r178336137) [](ancestors = 178336137)